### PR TITLE
HIVE-25008: Migrate hive table data into Iceberg format.

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?
Running ALTER TABLE ... SET TBLPROPERTIES('storage_handler'='org.apache.iceberg.mr.hive.HiveIcebergStorageHandler') should convert the data into iceberg format.
What is covered in this PR:

* Introduce new `commitAlterTable` hook
* update alterTable functions on HMSClient side to call the hook pre and commit alter table hook methods
* provide `preAlterTable` hook implementation on HiveIcebergMetaHook side, which does the following:
   * check for existing iceberg tables, and if the check fails it means that the iceberg table was not yet created
   * validate that the source hms table is external, non-bucketed, non-temporary and non-acid
   * collect schema, partitionspec, location, format metadata needed for data migration
   * if the hms table is partitioned, remove the partition keys from it, and add them as standard columns. Iceberg only have support for identity partitioned tables. 
   * update input/output format and serde with iceberg ones.
 * provide `commitAlterTable` hook implementation, which takes care of the iceberg table creation and data migration
 * create utility classes `HiveTableUtil` and `DataUtil`which collect the data files under the table/partition directories and updates the iceberg metadata files. Note: DataUtil is already marked as deprecated. The class is a copy of DataUtil from the upstream Iceberg project. Once the next release of Iceberg is available we should remove it. 
 * introduced new q tests

### Why are the changes needed?
This change would be useful when trying to read native hive table data through Iceberg (for ex: airlines data) 

### How was this patch tested?
Manual test, Q test

### CODE REVIEW

**Comment by lcspinter** [comment]
@pvary @marton-bod Could you please have a look at this change? Thanks

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+public class DataUtil {
+
+  private DataUtil() {
+  }
+
+  private static final PathFilter HIDDEN_PATH_FILTER =
+      p -> !p.getName().startsWith("_") && !p.getName().startsWith(".");
+
+
+    /**
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java:55]: nit:format

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import org.apache.iceberg.parquet.ParquetUtil;
+import org.apache.parquet.hadoop.ParquetFileReader;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+
+public class DataUtil {
+
+  private DataUtil() {
+  }
+
+  private static final PathFilter HIDDEN_PATH_FILTER =
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java:51]: Maybe? `org.apache.hadoop.hive.common.FileUtils.HIDDEN_FILES_PATH_FILTER`

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+import org.apache.iceberg.PartitionField;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.hadoop.HadoopInputFile;
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.orc.OrcMetrics;
+import org.apache.iceberg.parquet.ParquetUtil;
+import org.apache.parquet.hadoop.ParquetFileReader;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+
+public class DataUtil {
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java:46]: Is this the iceberg code we need here as well?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 4]
```diff
@@ -108,29 +113,14 @@ public void preCreateTable(org.apache.hadoop.hive.metastore.api.Table hmsTable)
 
     Schema schema = schema(catalogProperties, hmsTable);
     PartitionSpec spec = spec(schema, catalogProperties, hmsTable);
-
+    catalogProperties.put(InputFormatConfig.TABLE_SCHEMA, SchemaParser.toJson(schema));
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:116]: Why is this moved?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 5]
```diff
@@ -250,4 +312,11 @@ private static PartitionSpec spec(Schema schema, Properties properties,
       return PartitionSpec.unpartitioned();
     }
   }
+
+  private static class PreAlterTableProperties {
+    private static String tableLocation;
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:317]: Why do we do this? What happens if we have 2 parallel alterTable-s?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 6]
```diff
@@ -190,6 +180,78 @@ public void commitDropTable(org.apache.hadoop.hive.metastore.api.Table hmsTable,
     }
   }
 
+  @Override
+  public void preAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable, EnvironmentContext context)
+      throws MetaException {
+    HiveMetaHook.super.preAlterTable(hmsTable, context);
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:186]: any specific reason to do this?
Not wrong, just maybe we should do it everywhere or nowhere

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 7]
```diff
@@ -190,6 +180,78 @@ public void commitDropTable(org.apache.hadoop.hive.metastore.api.Table hmsTable,
...
+    HiveMetaHook.super.commitAlterTable(hmsTable, partitionSpecProxy);
+    if (generateMetadata) {
+      catalogProperties = getCatalogProperties(hmsTable);
+      catalogProperties.put(InputFormatConfig.TABLE_SCHEMA, SchemaParser.toJson(PreAlterTableProperties.schema));
+      catalogProperties.put(InputFormatConfig.PARTITION_SPEC, PartitionSpecParser.toJson(PreAlterTableProperties.spec));
+      if (Catalogs.hiveCatalog(conf)) {
+        catalogProperties.put(TableProperties.ENGINE_HIVE_ENABLED, true);
+      }
+      icebergTable = Catalogs.createTable(conf, catalogProperties);
+      HiveTableUtil.importFiles(PreAlterTableProperties.tableLocation, PreAlterTableProperties.format,
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:232]: What happens if this fails? Any chance to rollback?

Comment (pvary) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:232]: Or at least recover from the failure?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+import org.apache.iceberg.mapping.NameMappingParser;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HiveTableUtil {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HiveTableUtil.class);
+
+  private static class DataFileFilter implements PathFilter {
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:61]: This seems very similar to the HIDDEN_PATH_FILTER. Why don't we use that, or the Hive version of the filter?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        List<DataFile> dataFiles = getDataFiles(fileSystem, new Path(uri), Collections.emptyMap(), format, spec,
+            metricsConfig, nameMapping, conf);
+        dataFiles.forEach(append::appendFile);
+      } else {
+        PartitionSpecProxy.PartitionIterator iterator = partitionSpecProxy.getPartitionIterator();
+        List<Callable<List<DataFile>>> tasks = new ArrayList<>();
+        while (iterator.hasNext()) {
+          Partition partition = iterator.next();
+          Callable<List<DataFile>> task = () -> {
+            Path partitionPath = new Path(partition.getSd().getLocation());
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:110]: Using the location is not correct. For every partition the location could be set individually

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+            Map<String, String> partitionSpec = Warehouse.makeSpecFromName(partitionPath.toString());
+            return getDataFiles(fileSystem, partitionPath, partitionSpec, format.toLowerCase(), spec, metricsConfig,
+                nameMapping, conf);
+          };
+          tasks.add(task);
+        }
+        int numThreads = HiveConf.getIntVar(conf, HiveConf.ConfVars.HIVE_SERVER2_ICEBERG_METADATA_GENERATOR_THREADS);
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads,
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:118]: Shall we shut down the executor?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+          Partition partition = iterator.next();
+          Callable<List<DataFile>> task = () -> {
+            Path partitionPath = new Path(partition.getSd().getLocation());
+            Map<String, String> partitionSpec = Warehouse.makeSpecFromName(partitionPath.toString());
+            return getDataFiles(fileSystem, partitionPath, partitionSpec, format.toLowerCase(), spec, metricsConfig,
+                nameMapping, conf);
+          };
+          tasks.add(task);
+        }
+        int numThreads = HiveConf.getIntVar(conf, HiveConf.ConfVars.HIVE_SERVER2_ICEBERG_METADATA_GENERATOR_THREADS);
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:117]: Maybe not worth it, just wondering... Do we want to have a separate codepath where the threadnum is 1?

File: itests/qtest/pom.xml [hunk 12]
```diff
@@ -118,6 +118,11 @@
       <classifier>tests</classifier>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (pvary) [inline itests/qtest/pom.xml:121]: I think we should create our own module in itest, like `qtest-iceberg`

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+    /**
+     * Returns the data files in a partition by listing the partition location.
+     *
+     * For Parquet and ORC partitions, this will read metrics from the file footer. For Avro partitions,
+     * metrics are set to null.
+     *
+     * @param partition partition key, e.g., "a=1/b=2"
+     * @param uri partition location URI
+     * @param format partition format, avro or parquet
```
Comment (marton-bod) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java:63]: no `orc`?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+      p -> !p.getName().startsWith("_") && !p.getName().startsWith(".");
+
+
+    /**
+     * Returns the data files in a partition by listing the partition location.
+     *
+     * For Parquet and ORC partitions, this will read metrics from the file footer. For Avro partitions,
+     * metrics are set to null.
+     *
+     * @param partition partition key, e.g., "a=1/b=2"
```
Comment (marton-bod) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java:61]: maybe map of partition keys, since it's not singular?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+  private static List<DataFile> listAvroPartition(Map<String, String> partitionPath, String partitionUri,
+                                                  PartitionSpec spec, Configuration conf) {
+    try {
+      Path partition = new Path(partitionUri);
+      FileSystem fs = partition.getFileSystem(conf);
+      return Arrays.stream(fs.listStatus(partition, HIDDEN_PATH_FILTER))
+              .filter(FileStatus::isFile)
+              .map(stat -> {
+                Metrics metrics = new Metrics(0L, null, null, null);
```
Comment (marton-bod) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java:116]: Maybe a comment to indicate that this a workaround to set the record count to 0. Even though it's not correct to do so, there's no other way really

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 16]
```diff
@@ -190,6 +180,78 @@ public void commitDropTable(org.apache.hadoop.hive.metastore.api.Table hmsTable,
...
+    HiveMetaHook.super.preAlterTable(hmsTable, context);
+    catalogProperties = getCatalogProperties(hmsTable);
+    try {
+      icebergTable = Catalogs.loadTable(conf, catalogProperties);
+    } catch (NoSuchTableException nte) {
+      // If the iceberg table does not exist, and the hms table is external and not temporary, not acid,
+      // and not bucketed, we will create it in commitAlterTable
+      StorageDescriptor sd = hmsTable.getSd();
+      generateMetadata = MetaStoreUtils.isExternalTable(hmsTable) && !hmsTable.isTemporary() &&
+          !(sd.getNumBuckets() > 0) && !AcidUtils.isTransactionalTable(hmsTable);
```
Comment (marton-bod) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:195]: Can we not migrate a bucketed table to Iceberg? Would the table just lose its bucketed property but otherwise the migration would succeed, or do we run into some errors because of the way the execution plan is put together differently for bucketed tables?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 17]
```diff
@@ -66,6 +74,7 @@
   private boolean deleteIcebergTable;
   private FileIO deleteIo;
   private TableMetadata deleteMetadata;
+  private boolean generateMetadata;
```
Comment (marton-bod) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:77]: How about a different naming that reflects that it's related to "migratability"? e.g. `boolean canMigrateHiveTable`

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 5 above)
Comment (marton-bod) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:317]: Why make the fields static? I also think it could create concurrency issues. You could create this as an object similarly to how the catalogProperties are currently cached in the MetaHook between pre and post executions

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads,
+            new ThreadFactoryBuilder().setNameFormat("Iceberg metadata generator").setDaemon(true).build());
```
Comment (marton-bod) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:119]: nit: I think we usually use a format like this for threadpool naming: `iceberg-metadata-generator-%d`

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+        List<Future<List<DataFile>>> futures = executor.invokeAll(tasks);
+        for (Future<List<DataFile>> future : futures) {
+          List<DataFile> dataFiles = future.get();
+          dataFiles.forEach(append::appendFile);
+        }
+      }
+
+    } catch (URISyntaxException | IOException | InterruptedException | ExecutionException e) {
+      LOG.error("Cannot import hive data into iceberg table", e);
```
Comment (marton-bod) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:128]: Are we sure want to swallow the exception here? If the import failed or only partially succeeded, we're still committing the appends below and not fail the operation. Couldn't this lead to an inconsistent state and silently hiding failures from the user?

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java (hunk 2 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/DataUtil.java:51]: This class is a copy of org.apache.iceberg.data.DataUtil from upstream Iceberg project. I would rather not change it, since it will be completely removed once Iceberg 0.12 is released.

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 6 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:186]: The default implementation of preAlterTable does some basic validation of the alter operation type. I thought this validation is useful.

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 5 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:317]: Frankly, I don't know why I left it as static :). Good catch!

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java (hunk 8 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:61]: This filter also skips directories. We need additional condition because when we create the iceberg table, it creates the metadata directory, and we don't want to scan that for data files.

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java (hunk 9 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:110]: Yes, you're right. Fixed it.

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java (hunk 18 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:119]: Fixed.

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java (hunk 19 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:128]: You are correct. I refactored the code to get back with an exception.

File: itests/qtest/pom.xml (hunk 12 above)
Comment (lcspinter) [reply itests/qtest/pom.xml:121]: I opened an upstream [jira](https://issues.apache.org/jira/browse/HIVE-25010) to track this work.

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 16 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:195]: It turns out we can :). Removed this condition.

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java (hunk 8 above)
Comment (pvary) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:59]: I have concerns with this:
- AFAIK Hive allows directories under a table, and scans them for files for normal tables. So we might want to handle directories.
- `fileSystem.isFile(path)` is an extra call for every file. This could be very-very costly for big tables. Any way to avoid the call?

Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:59]: As a workaround, I could substitute the `isFile()` call with a `!name.equals("metadata")`, but with this, we might have issues when we have the data files under some metadata folder.

Comment (pvary) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java:59]: Is this only called to a non-Iceberg / external Hive table only? Do we have to be scared to metadata folders?

Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveTableUtil.java]: Moved the file listing before the iceberg table is created, therefore the contents of the metadata directory are not messing up the list of data files.

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java (hunk 7 above)
Comment (lcspinter) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java]: As discussed, I will create a new PR the tackle the rollback.

**Review by pvary — APPROVED** [review]

**Comment by lcspinter** [comment]
Thanks, @marton-bod  and @pvary for the review!
